### DISCUSSION TITLE
Logging training to external file

### DISCUSSION THREAD

**Discussion by lluissalord** [explosion/spaCy#8356] [Help: Coding & Implementations] [answered]
Hi all,

I am trying to see how could I log all the information printed out by the `ConsoleLogger` into an external file.

I imagine that I could reimplement the `ConsoleLogger` with a different `stdout`, but I think that the best way would be to reuse the already created function `console_logger` (shown below) but using different `stdout` and `stderr` parameters.

https://github.com/explosion/spaCy/blob/ed561cf428494c2b7a6790cd4b91b5326102b59d/spacy/training/loggers.py#L27-L33

Could you help me finding out how should I do it?

Thanks!

**Comment by svlandeg** [comment, marked as answer]
The `spacy train` CLI command refers to `sys.stdout` and `sys.stderr` for these, so I think you should be able to just set these from within your script, or alternatively on the command line just redirect the output to a file?

See also https://stackoverflow.com/questions/4675728/redirect-stdout-to-a-file-in-python

Alternatively, have a look at the docs [here](https://spacy.io/usage/training#custom-logging) that outline how to implement your own custom file logger!

**Reply by lluissalord** [reply]
What do you mean to set these within my script?

To put some context, currently I am launching the CLI command using `os.system` function on Python (as this is the language used on the whole project).

Regarding redirecting the output, I tried the following code but it did not worked:
```python
        try:
            with open('train_logging.log', 'w') as f:
                sys.stdout = f
                os.system(f'python -m spacy train {self.config_filepath} --output {self.output_path} --paths.train {self.train_filepath} --paths.dev {self.dev_filepath}')
        finally:
            sys.stdout = sys.__stdout__
```

**Reply by svlandeg** [reply]
You can find some tips on working with `os.system` here: https://stackoverflow.com/questions/3197509/redirecting-stdio-from-a-command-in-os-system-in-python

These are general-purpose Python solutions to redirect/change the `stdout` and `stderr`. If you want to have more control on the way spaCy logs things, a custom logger (cf link cited in previous answer) will probably be the best solution for you.

**Reply by lluissalord** [reply]
Thanks @svlandeg I will continue with some of your suggestions.
